clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java

```diff
@@ -121,7 +121,12 @@ public void put(long key, Chain chain) {
     try {
       segmentFor(key).put(key, chain);
     } catch (OversizeMappingException e) {
-      handleOversizeMappingException(key, (long k) -> segmentFor(k).put(k, chain));
+      try {
+        handleOversizeMappingException(key, (long k) -> segmentFor(k).put(k, chain));
+      } catch (OversizeMappingException e1) {
+        segmentFor(key).remove(key);
+        throw e;
```

Comment:
This is ugly but ensures that key is removed only in put. Do we need this catch black around other handleOversizeMappingException too? If yes, the method itself can be refactored.

Reply:
Wouldn't it be easier to do:
```java
try {
  try {
    segmentFor(key).put(key, chain);
  } catch (OversizeMappingException e) {
    handleOversizeMappingException(key, (long k) -> segmentFor(k).put(k, chain));
  }
} catch (Throwable t) {
  segmentFor(key).remove(key);
  throw t;
}
```
Thats the general idea right... the fact that we trap the OSME as harmless further out is an independent thing. The general idea is "something went wrong when putting" not sure what happened, safest thing to do is invalidate.